advanced/Scripts/dnsmasq_parse.py

```diff
@@ -0,0 +1,179 @@
+q_re = re.compile(r'(.*) dnsmasq\[\d+\]: query\[(.*)\] (.*) from (.*)')
+f_re = re.compile(r'(.*) dnsmasq\[\d+\]: forwarded (.*) to (.*)')
+r_re = re.compile(r'(.*) dnsmasq\[\d+\]: (reply|cached) (.*) is (.*)')
+
+
+#-----------------------------------------------------------------------------
+# Functions
+#-----------------------------------------------------------------------------
+def create_tables():
+    qt = '''
+    CREATE TABLE IF NOT EXISTS queries (
+        source text,
+        query_type text,
+        name text,
+        ts datetime,
+        piholed bool
+    )
+    '''
+    c.execute(qt)
+    conn.commit()
+
+    ft = '''
+    CREATE TABLE IF NOT EXISTS forwards (
+        resolver text,
+        name text,
+        ts datetime
+    )
+    '''
+    c.execute(ft)
+    conn.commit()
+
+    rt = '''
+    CREATE TABLE IF NOT EXISTS replies (
+        ip text,
+        reply_type text,
+        name text,
+        ts datetime
+    )
+    '''
+    c.execute(rt)
+    conn.commit()
+
+
+def convert_date(ds):
+    y = str(datetime.datetime.now().year)
+    ltime = time.strptime('{0} {1}'.format(y, ds), '%Y %b %d %H:%M:%S')
+
+    return time.strftime('%Y-%m-%d %H:%M:%S', ltime)
+
+
+def parse_query(query):
+    m = q_re.match(query)
+    if m is not None:
+        add_query(m.group(4), m.group(2), m.group(3), m.group(1),check_gravity(m.group(3)))
+
+
+def check_gravity(txt):
+    for entry in gravFile:
+        if entry == txt:
+            return True
+    return False
+
+
+def parse_forward(query):
+    m = f_re.match(query)
+    if m is not None:
+        add_forward(m.group(3), m.group(2), m.group(1))
+
+
+def parse_reply(query):
+    m = r_re.match(query)
+    if m is not None:
+        add_reply(m.group(4), m.group(2), m.group(3), m.group(1))
+
+
+def add_query(source, qtype, name, ts, blocked):
+    sql = "INSERT INTO queries (source, query_type, name, ts, piholed) VALUES(?,?,?,?,?)"
+    c.execute(sql, (source, qtype, name, convert_date(ts), blocked))
+
+
+def add_forward(resolver, name, ts):
+    sql = "INSERT INTO forwards (resolver, name, ts) VALUES(?,?,?)"
+    c.execute(sql, (resolver, name, convert_date(ts)))
+
+
+def add_reply(ip, rtype, name, ts):
+    sql = "INSERT INTO replies (ip, reply_type, name, ts) VALUES(?,?,?,?)"
+    c.execute(sql, (ip, rtype, name, convert_date(ts)))
+
+
+#-----------------------------------------------------------------------------
+# Main
+#-----------------------------------------------------------------------------
+if len(sys.argv) != 2:
+    print 'Usage: dnsmasq_parse.py logfile'
+    sys.exit()
+
+logfile = sys.argv[1]
+
+# Create the SQLite connection
+conn = sqlite3.connect('/etc/pihole/pihole.db')
+
+gravFile = []
+
+with open('/etc/pihole/gravity.list', 'r') as g:
+    for line in g:
+        gravFile.append(line.rstrip().split(' ')[1])
+
+with conn:
+    c = conn.cursor()
+
+    create_tables()
+
+    # Parse the log file.
+    with open(logfile) as f:
+        for line in f:
```

Comment:
using re.compile before the loop would be more precise and possibly perform better.